/**
 * Copyright Soramitsu Co., Ltd. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

syntax = "proto3";
package iroha.protocol;
import "block.proto";
import "transaction.proto";
import "primitive.proto";

// *** WSV data structure *** //
message Asset {
  string asset_id = 1;
  string domain_id = 2;
  uint32 precision = 3;
}

message Domain {
  string domain_id = 1;
  string default_role = 2;
}

message Account {
  string account_id = 1;
  string domain_id = 2;
  uint32 quorum = 3;
  string json_data = 4;
}

message AccountAsset {
  string asset_id = 1;
  string account_id = 2;
  string balance = 3;
}

// *** Responses *** //
message AccountAssetResponse {
  repeated AccountAsset account_assets = 1;
}

message AccountDetailResponse {
  string detail = 1;
}

message AccountResponse {
  Account account = 1;
  repeated string account_roles = 2;
}

message AssetResponse {
  Asset asset = 1;
}

message RolesResponse {
  repeated string roles = 1;
}

message RolePermissionsResponse {
  repeated RolePermission permissions = 1;
}

message ErrorResponse {
  enum Reason {
    STATELESS_INVALID = 0;
    STATEFUL_INVALID = 1;
    NO_ACCOUNT = 2;         // when requested account does not exist
    NO_ACCOUNT_ASSETS = 3;  // when requested account asset does not exist
    NO_ACCOUNT_DETAIL = 4;  // when requested account detail does not exist
    NO_SIGNATORIES = 5;     // when requested signatories does not exist
    NOT_SUPPORTED = 6;      // when unidentified request was received
    NO_ASSET = 7;           // when requested asset does not exist
    NO_ROLES = 8;           // when there are no roles defined in the system
  }
  Reason reason = 1;
  string message = 2;
}

message SignatoriesResponse {
  repeated bytes keys = 1;
}

message TransactionsResponse {
  repeated Transaction transactions = 1;
}

message QueryResponse {
  oneof response {
    AccountAssetResponse account_assets_response = 1;
    AccountDetailResponse account_detail_response = 2;
    AccountResponse account_response = 3;
    ErrorResponse error_response = 4;
    SignatoriesResponse signatories_response = 5;
    TransactionsResponse transactions_response = 6;
    AssetResponse asset_response = 7;
    RolesResponse roles_response = 8;
    RolePermissionsResponse role_permissions_response = 9;
  }
  string query_hash = 10;
}

message BlockResponse {
  Block block = 1;
}

message BlockErrorResponse {
  string message = 1;
}

message BlockQueryResponse {
  oneof response {
    BlockResponse block_response = 1;
    BlockErrorResponse error_response = 2;
  }
}
